hive/cell: remove health reporting on health provider.

tommyp1ckles opened the pull request:
Restructures health provider (i.e cell.Health implementations) to accept opaque Update interface type instead
of string.

As well, this removes health reporting functionality from the cell.Health provider implementations and moves all health reporting to the scoped reporter in cell/structured.go.

Previously, with https://github.com/cilium/cilium/pull/27522, the structured health reporter in cell/structured.go would wrap a passed cell.HealthReporter which was implemented *healthReporter which was given out by *healthProvider (cell.Health).

This made reporting from the structured reporting awkward as it forced it to pass a string, so the data from the structured reporter had to be flattened into a string before being updated for the modules status in the health provider.

The cell.Health implementation in cell/health.go now serves to just process and store updates from the structured reporter, as well as providing subscription functionality.

As a follow up to this PR, I will use this to change the modular health OpenAPI response model to include the [full structured output from the health reporter.](https://github.com/cilium/cilium/pull/28804/files#diff-508a09ed0e66f45b2f1c78e3650eedf4de551b63150c6dac457298af46f9152bR1388)

**The goal of this PR is to:**
* Remove the old health reporter that's not necessary anymore.
* Store structured data in HealthProvider so we can send a non-flattened version of the report via the "cilium status" command.
* Add metrics for number of ok/degraded/unknown modules

tommyp1ckles committed:
cilium-debug: don't print modules with unknown health status.

Modules that have an unknown status (i.e. they take a health reporter
scope but never report an initial status) should be ignored as to not
create confusing output.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
hive: add hive_health_status_levels metric.

This provides a count of modules registered in the health provider, by
health status level.

Example output:

```
cilium_hive_status                                       status="OK"                                                                                                                                                                                                                                                      2.000000
cilium_hive_status                                       status="Unknown"                                                                                                                                                                                                                                                 1.000000
```

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles commented on pkg/health/client/modules.go:
```diff
@@ -26,18 +26,22 @@ func GetAndFormatModulesHealth(w io.Writer, clt ModulesHealth, verbose bool) {
-				fmt.Fprintf(w, "  %s\t%s\t%12s\n%s", m.ModuleID, m.Level, m.LastUpdated, m.Message)
-			} else {
-				fmt.Fprintf(w, "  %s\t%s\t%s\t%12s\n", m.ModuleID, m.Level, m.Message, m.LastUpdated)
+			n := &cell.StatusNode{}
```
**Note**: This is a temporary measure - following this PR we will be able to change the OpenAPI model to pass the [entire update structure](https://github.com/cilium/cilium/pull/28804/files#diff-508a09ed0e66f45b2f1c78e3650eedf4de551b63150c6dac457298af46f9152bR1388) instead of sending the data as a json string.

derailed replied:
I am actually not sure about this? Thinking this data will eventually be stored in state db. Likely this would results in a collection of rows when queried and it will be up to the call site to render a tree if needed?

joamaki replied:
Related: just added REST API for statedb to query tables which would allow transferring the `StatusNode`s (gob encoded) over without having to do the whole openapi.yaml dance: https://github.com/cilium/cilium/pull/28872. 
The hierarchical structure would need to be reconstructed since it's just a flat collection of nodes. `ID` plus `ParentID` would probably do it. Maybe too soon to think about statedb use for this, but perhaps could already design the data model to suite it.

tommyp1ckles replied:
@joamaki this makes sense, as mentioned I have a follow up to this PR where I was going to change the openapi spec to include the tree structure but I think i'll change that to have a flat list of nodes/id/parent-ids to be close to an eventual statedb schema.

tommyp1ckles replied:
Re: discussion points with @derailed regarding statedb, discussing this offline we've compromised to continue cleaning up current iteration of modular health reporting and begin looking at possibly transitioning to statedb following the next release.

Will resolve

danehans commented on pkg/health/client/modules_test.go:
```diff
@@ -30,14 +30,6 @@ func TestGetAndFormatModulesHealth(t *testing.T) {
-  m1	OK	a ok	          2s
-  m2	Degraded	doh	         20s`,
-			v: true,
-		},
```
Is this being removed due to https://github.com/cilium/cilium/pull/28773/files#r1372564148?

tommyp1ckles replied:
Yes, as well as some work being done on making the CLI cleaner: https://github.com/cilium/cilium/pull/28800

However, removing this is probably not the way to go so I've re-added the test case in my last commit with some changes.

danehans commented on pkg/hive/cell/health.go:
```diff
@@ -57,6 +58,25 @@ type HealthReporter interface {
 	Degraded(reason string, err error)
 }
 
+// Update represents a instantaneous health status update.
```
nit: s/a/an/

danehans reviewed (commented):
Leaving a few comments. I'm not familiar enough with agent health reporting to give this a proper review. It might be best to add another reviewer to give this PR the proper treatment.

tommyp1ckles committed:
hive/cell: remove health reporting on health provider.

Restructures health provider (i.e cell.Health implementations) to accept opaque Update interface type instead
of string.

This removes health reporting functionality from the cell.Health provider implementations and moves all health reporting to
the scoped reporter in cell/structured.go.

Previously, the structured health reporter in cell/structured.go would wrap a passed cell.HealthReporter which was implemented *healthReporter which was given out by *healthProvider (cell.Health).

This made reporting from the structured reporting awkward as it forced
it to pass a string, so the data from the structured reporter had to be
flattened into a string before being updated for the modules status in
the health provider.

The cell.Health implementation in cell/health.go now serves to just
process and store updates from the structured reporter, as well as
providing subscription functionality (currently just used by tests - but
will be used for metrics in subsequent commits).

Stored "Status's" are also now an interface type that allows for more flexibility when getting reported status.

This allows for sending a json representation of the structured data via
the rest api endpoint.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
pkg/cell: use cell.StatusNode to render output and fix test.

Previous commits changed how health data is sent, where the health
status CLI dbg command now json unmarshals into *cell.StatusNode.

With this, we can simplify writing of health status data by moving all
text rendering logic into (*cell.StatusNode).StringIndent(...).

As well, drop unnecessary column headers from CLI output and fix tests to reflect this.

Example of new output:

```
Modules Health:
  agent.controlplane.endpoint-manager
   endpoint-gc OK endpoint-gc 6m31s ago (x2)
   cilium-endpoint-2921 (/)
    datapath-regenerate OK Endpoint regeneration successful 6m25s ago (x1)
    policy map sync     OK sync-policymap-2921              6m24s ago (x1)
   cilium-endpoint-4029 (/)
    policy map sync     OK sync-policymap-4029              6m21s ago (x1)
    datapath-regenerate OK Endpoint regeneration successful 6m24s ago (x1)
  agent.controlplane.node-manager
   background-sync OK Node validation successful 6m31s ago (x7
```

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

joamaki commented on pkg/hive/cell/health.go:
```diff
@@ -66,42 +86,67 @@ type Health interface {
 	// Stop stops the health provider from processing updates.
 	Stop(context.Context) error
 
+	// Subscribe to health status updates.:w
```
`:w` :)

derailed commented on pkg/hive/cell/health.go:
```diff
@@ -66,42 +86,70 @@ type Health interface {
-type Update struct {
-	Level
+// UpdateTemp is an event that denotes the change of a modules health state.
+type UpdateTemp struct {
```
Where is this used?

tommyp1ckles replied:
woops - leftover code, ill remove.

derailed commented on pkg/hive/cell/health.go:
```diff
@@ -57,6 +58,25 @@ type HealthReporter interface {
+
+	// JSON returns a JSON representation of the update, this is used by the agent
+	// health CLI to unmarshal health status into cell.StatusNode.
+	JSON() ([]byte, error)
```
Should we mix in json.Marshaller instead?

tommyp1ckles replied:
I tried that, the UnmarshalJSON function on recursive data structure tends to cause problems 😉

derailed commented on pkg/hive/cell/health.go:
```diff
@@ -66,42 +86,70 @@ type Health interface {
 // Status is a modules last health state, including the last update.
 type Status struct {
-	// Update is the last reported update for a module.
+	// UpdateTemp is the last reported update for a module.
```
Per above^ incorrect go doc?

derailed commented on pkg/hive/cell/health.go:
```diff
@@ -110,32 +158,49 @@ func (s *Status) String() string {
 	} else {
 		sinceLast = time.Since(s.LastUpdated).String() + " ago"
```
This seems to be a rendering decision? Perhaps we should let consumers defined how since should be formatted.

tommyp1ckles replied:
Yeah I agree, but let's try to get all the formatting in one place, and eventually move this code elsewhere (i.e. we can implement different formatting modes via aliases on *UpdateNode).

derailed reviewed (changes requested):
@tommyp1ckles Nice work! 
Thinking eventually this data will reside in stateDB, I am not sure marshaling and storing as tree is correct for the future. I was hoping we could just return a collection ModuleStatus and let the call site hydrate as a tree if needed??
